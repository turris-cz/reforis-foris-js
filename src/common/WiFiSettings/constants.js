/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

export const HTMODES = {
    NOHT: _("Disabled"),
    HT20: _("802.11n - 20 MHz wide channel"),
    HT40: _("802.11n - 40 MHz wide channel"),
    VHT20: _("802.11ac - 20 MHz wide channel"),
    VHT40: _("802.11ac - 40 MHz wide channel"),
    VHT80: _("802.11ac - 80 MHz wide channel"),
    VHT80_80: _("802.11ac - 80+80 MHz wide channel"),
    VHT160: _("802.11ac - 160 MHz wide channel"),
    HE20: _("802.11ax - 20 MHz wide channel"),
    HE40: _("802.11ax - 40 MHz wide channel"),
    HE80: _("802.11ax - 80 MHz wide channel"),
    HE80_80: _("802.11ax - 80+80 MHz wide channel"),
    HE160: _("802.11ax - 160 MHz wide channel"),
};
export const BANDS = {
    "2g": "2.4",
    "5g": "5",
    "6g": "6",
};
export const ENCRYPTIONMODES = {
    WPA3: _("WPA3 only"),
    "WPA2/3": _("WPA3 with WPA2 as fallback (default)"),
    WPA2: _("WPA2 only"),
};
export const HELP_TEXTS = {
    ssid: _(
        "SSID which contains non-standard characters could cause problems on some devices."
    ),
    password: _(
        "WPA2/3 pre-shared key, that is required to connect to the network."
    ),
    hidden: _(
        "If set, network is not visible when scanning for available networks."
    ),
    band: _(
        "The 2.4 GHz band is more widely supported by clients, but tends to have more interference. The 5 GHz band is a newer standard and may not be supported by all your devices. It usually has less interference, but the signal does not carry so well indoors."
    ),
    htmode: _(
        "Change this to adjust 802.11n/ac/ax mode of operation. 802.11n with 40 MHz wide channels can yield higher throughput but can cause more interference in the network. If you don't know what to choose, use the default option with 20 MHz wide channel."
    ),
    guest_wifi_enabled: _(
        "Enables Wi-Fi for guests, which is separated from LAN network. Devices connected to this network are allowed to access the internet, but aren't allowed to access other devices and the configuration interface of the router. Parameters of the guest network can be set in the Guest network tab."
    ),
    wpa3: _(
        "The WPA3 standard is the new most secure encryption method that is suggested to be used with any device that supports it. The older devices without WPA3 support require older WPA2. If you experience issues with connecting older devices, try to enable WPA2."
    ),
};
