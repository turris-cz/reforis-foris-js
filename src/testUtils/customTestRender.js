/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from "react";
import { UIDReset } from "react-uid";
import { StaticRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import PropTypes from "prop-types";

import { AlertContextMock } from "./alertContextMock";
import { CustomizationContextMock } from "./customizationContextMock";

Wrapper.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};

function Wrapper({ children }) {
    return (
        <CustomizationContextMock>
            <AlertContextMock>
                <StaticRouter>
                    <UIDReset>{children}</UIDReset>
                </StaticRouter>
            </AlertContextMock>
        </CustomizationContextMock>
    );
}

const customTestRender = (ui, options) =>
    render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customTestRender as render };
