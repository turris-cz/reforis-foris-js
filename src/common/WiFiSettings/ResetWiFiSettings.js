/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useState } from "react";

import PropTypes from "prop-types";

import { useAPIPost } from "../../api/hooks";
import { API_STATE } from "../../api/utils";
import { ALERT_TYPES } from "../../bootstrap/Alert";
import Button from "../../bootstrap/Button";
import { formFieldsSize } from "../../bootstrap/constants";
import { useAlert } from "../../context/alertContext/AlertContext";

ResetWiFiSettings.propTypes = {
    ws: PropTypes.object.isRequired,
    endpoint: PropTypes.string.isRequired,
};

function ResetWiFiSettings({ ws, endpoint }) {
    const [isLoading, setIsLoading] = useState(false);

    useEffect(() => {
        const module = "wifi";
        ws.subscribe(module).bind(module, "reset", () => {
            // eslint-disable-next-line no-restricted-globals
            setTimeout(() => location.reload(), 1000);
        });
    }, [ws]);

    const [postResetResponse, postReset] = useAPIPost(endpoint);
    const [setAlert, dismissAlert] = useAlert();
    useEffect(() => {
        if (postResetResponse.state === API_STATE.ERROR) {
            setAlert(_("An error occurred during resetting Wi-Fi settings."));
        } else if (postResetResponse.state === API_STATE.SUCCESS) {
            setAlert(
                _("Wi-Fi settings are set to defaults."),
                ALERT_TYPES.SUCCESS
            );
        }
    }, [postResetResponse, setAlert]);

    const onReset = () => {
        dismissAlert();
        setIsLoading(true);
        postReset();
    };

    return (
        <div className={formFieldsSize}>
            <h2>{_("Reset Wi-Fi Settings")}</h2>
            <p>
                {_(
                    "If a number of wireless cards doesn't match, you may try to reset the Wi-Fi settings. Note that this will remove the current Wi-Fi configuration and restore the default values."
                )}
            </p>
            <div className="text-end">
                <Button
                    className="btn-primary"
                    forisFormSize
                    loading={isLoading}
                    disabled={isLoading}
                    onClick={onReset}
                >
                    {_("Reset Wi-Fi Settings")}
                </Button>
            </div>
        </div>
    );
}

export default ResetWiFiSettings;
