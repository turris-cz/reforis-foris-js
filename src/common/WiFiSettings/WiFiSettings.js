/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import PropTypes from "prop-types";

import ResetWiFiSettings from "./ResetWiFiSettings";
import WiFiForm from "./WiFiForm";
import ForisForm from "../../form/components/ForisForm";

WiFiSettings.propTypes = {
    ws: PropTypes.object.isRequired,
    endpoint: PropTypes.string.isRequired,
    resetEndpoint: PropTypes.string.isRequired,
    hasGuestNetwork: PropTypes.bool,
};

function WiFiSettings({ ws, endpoint, resetEndpoint, hasGuestNetwork }) {
    return (
        <>
            <ForisForm
                ws={ws}
                forisConfig={{
                    endpoint,
                    wsModule: "wifi",
                }}
                prepData={prepData}
                prepDataToSubmit={prepDataToSubmit}
                validator={validator}
            >
                <WiFiForm hasGuestNetwork={hasGuestNetwork} />
            </ForisForm>
            <ResetWiFiSettings ws={ws} endpoint={resetEndpoint} />
        </>
    );
}

function prepData(formData) {
    formData.devices.forEach((device, idx) => {
        formData.devices[idx].channel = device.channel.toString();
    });
    return formData;
}

function prepDataToSubmit(formData) {
    formData.devices.forEach((device, idx) => {
        delete device.available_bands;

        formData.devices[idx].channel = parseInt(device.channel);

        if (!device.enabled) {
            formData.devices[idx] = { id: device.id, enabled: false };
            return;
        }

        if (!device.guest_wifi.enabled)
            formData.devices[idx].guest_wifi = { enabled: false };

        if (device.encryption === "WPA2") {
            delete formData.devices[idx].ieee80211w_disabled;
        }
    });
    return formData;
}

export function byteCount(string) {
    const buffer = Buffer.from(string, "utf-8");
    const count = buffer.byteLength;
    return count;
}

export function validator(formData) {
    const formErrors = formData.devices.map((device) => {
        if (!device.enabled) return {};

        const errors = {};
        if (device.SSID.length > 32)
            errors.SSID = _("SSID can't be longer than 32 symbols");
        if (device.SSID.length === 0) errors.SSID = _("SSID can't be empty");
        if (byteCount(device.SSID) > 32)
            errors.SSID = _("SSID can't be longer than 32 bytes");

        if (device.password.length < 8)
            errors.password = _("Password must contain at least 8 symbols");
        if (device.password.length >= 64)
            errors.password = _(
                "Password must not contain more than 63 symbols"
            );

        if (!device.guest_wifi.enabled) return errors;

        const guest_wifi_errors = {};
        if (device.guest_wifi.SSID.length > 32)
            guest_wifi_errors.SSID = _("SSID can't be longer than 32 symbols");
        if (device.guest_wifi.SSID.length === 0)
            guest_wifi_errors.SSID = _("SSID can't be empty");
        if (byteCount(device.guest_wifi.SSID) > 32)
            guest_wifi_errors.SSID = _("SSID can't be longer than 32 bytes");

        if (device.guest_wifi.password.length < 8)
            guest_wifi_errors.password = _(
                "Password must contain at least 8 symbols"
            );
        if (device.guest_wifi.password.length >= 64)
            guest_wifi_errors.password = _(
                "Password must not contain more than 63 symbols"
            );

        if (guest_wifi_errors.SSID || guest_wifi_errors.password) {
            errors.guest_wifi = guest_wifi_errors;
        }
        return errors;
    });
    return JSON.stringify(formErrors).match(/\[[{},?]+\]/) ? null : formErrors;
}

export default WiFiSettings;
