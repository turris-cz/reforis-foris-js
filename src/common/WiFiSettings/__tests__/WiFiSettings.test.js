/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import diffSnapshot from "snapshot-diff";
import mockAxios from "jest-mock-axios";

import { fireEvent, render, wait } from "customTestRender";
import WebSockets from "webSockets/WebSockets";
import { mockJSONError } from "testUtils/network";

import {
    wifiSettingsFixture,
    oneDevice,
    twoDevices,
    threeDevices,
} from "./__fixtures__/wifiSettings";
import WiFiSettings, { validator, byteCount } from "../WiFiSettings";

describe("<WiFiSettings/>", () => {
    let firstRender;
    let getAllByText;
    let getAllByLabelText;
    let getByText;
    let getByLabelText;
    let asFragment;
    const endpoint = "/reforis/api/wifi";

    beforeEach(async () => {
        const webSockets = new WebSockets();
        const renderRes = render(
            <WiFiSettings
                ws={webSockets}
                endpoint={endpoint}
                resetEndpoint="foo"
            />
        );
        asFragment = renderRes.asFragment;
        getAllByText = renderRes.getAllByText;
        getAllByLabelText = renderRes.getAllByLabelText;
        getByLabelText = renderRes.getByLabelText;
        getByText = renderRes.getByText;
        mockAxios.mockResponse({ data: wifiSettingsFixture() });
        await wait(() => renderRes.getByText("Wi-Fi 1"));
        firstRender = renderRes.asFragment();
    });

    it("should handle error", async () => {
        const webSockets = new WebSockets();
        const { getByText } = render(
            <WiFiSettings
                ws={webSockets}
                endpoint={endpoint}
                resetEndpoint="foo"
            />
        );
        const errorMessage = "An API error occurred.";
        mockJSONError(errorMessage);
        await wait(() => {
            expect(getByText(errorMessage)).toBeTruthy();
        });
    });

    it("Snapshot both modules disabled.", () => {
        expect(firstRender).toMatchSnapshot();
    });

    it("Snapshot one module enabled.", () => {
        fireEvent.click(getByText("Wi-Fi 1"));
        expect(diffSnapshot(firstRender, asFragment())).toMatchSnapshot();
    });

    it("Snapshot 2.4 GHz", () => {
        fireEvent.click(getByText("Wi-Fi 1"));
        const enabledRender = asFragment();
        fireEvent.click(getAllByText(/2.4/)[0]);
        expect(diffSnapshot(enabledRender, asFragment())).toMatchSnapshot();
    });

    it("Snapshot guest network.", () => {
        fireEvent.click(getByText("Wi-Fi 1"));
        const enabledRender = asFragment();
        fireEvent.click(getAllByText("Enable Guest Wi-Fi")[0]);
        expect(diffSnapshot(enabledRender, asFragment())).toMatchSnapshot();
    });

    it("Post form: both modules disabled.", () => {
        fireEvent.click(getByText("Save"));
        expect(mockAxios.post).toBeCalled();
        const data = {
            devices: [
                { enabled: false, id: 0 },
                { enabled: false, id: 1 },
            ],
        };
        expect(mockAxios.post).toHaveBeenCalledWith(
            endpoint,
            data,
            expect.anything()
        );
    });

    it("Post form: one module enabled.", () => {
        fireEvent.click(getByText("Wi-Fi 1"));

        fireEvent.click(getByText("Save"));
        expect(mockAxios.post).toBeCalled();
        const data = {
            devices: [
                {
                    SSID: "TestSSID1",
                    channel: 60,
                    enabled: true,
                    guest_wifi: { enabled: false },
                    hidden: false,
                    htmode: "HT80",
                    band: "5g",
                    id: 0,
                    password: "TestPass",
                    encryption: "WPA3",
                },
                { enabled: false, id: 1 },
            ],
        };
        expect(mockAxios.post).toHaveBeenCalledWith(
            endpoint,
            data,
            expect.anything()
        );
    });

    it("Post form: 2.4 GHz", () => {
        fireEvent.click(getByText("Wi-Fi 1"));
        fireEvent.click(getAllByText(/2.4/)[0]);

        fireEvent.click(getByText("Save"));
        expect(mockAxios.post).toBeCalled();
        const data = {
            devices: [
                {
                    SSID: "TestSSID1",
                    channel: 0,
                    enabled: true,
                    guest_wifi: { enabled: false },
                    hidden: false,
                    htmode: "VHT80",
                    band: "2g",
                    id: 0,
                    password: "TestPass",
                    encryption: "WPA3",
                },
                { enabled: false, id: 1 },
            ],
        };
        expect(mockAxios.post).toHaveBeenCalledWith(
            endpoint,
            data,
            expect.anything()
        );
    });

    it("Post form: guest network.", () => {
        fireEvent.click(getByText("Wi-Fi 1"));
        fireEvent.click(getAllByText("Enable Guest Wi-Fi")[0]);
        fireEvent.change(getAllByLabelText("Password")[1], {
            target: { value: "test_password" },
        });

        fireEvent.click(getByText("Save"));
        expect(mockAxios.post).toBeCalled();
        const data = {
            devices: [
                {
                    SSID: "TestSSID1",
                    channel: 60,
                    enabled: true,
                    guest_wifi: {
                        SSID: "TestGuestSSID",
                        enabled: true,
                        password: "test_password",
                    },
                    hidden: false,
                    htmode: "HT80",
                    band: "5g",
                    id: 0,
                    password: "TestPass",
                    encryption: "WPA3",
                },
                { enabled: false, id: 1 },
            ],
        };
        expect(mockAxios.post).toHaveBeenCalledWith(
            endpoint,
            data,
            expect.anything()
        );
    });

    it("Validator function using regex for one device", () => {
        expect(validator(oneDevice)).toEqual(null);
    });

    it("Validator function using regex for two devices", () => {
        const twoDevicesFormErrors = [{ SSID: "SSID can't be empty" }, {}];
        expect(validator(twoDevices)).toEqual(twoDevicesFormErrors);
    });

    it("Validator function using regex for three devices", () => {
        const threeDevicesFormErrors = [
            {},
            {},
            { password: "Password must contain at least 8 symbols" },
        ];
        expect(validator(threeDevices)).toEqual(threeDevicesFormErrors);
    });

    it("ByteCount function", () => {
        expect(byteCount("abc")).toEqual(3);
    });

    it("Should validate password length", () => {
        const shortErrorFeedback = /Password must contain/i;
        const longErrorFeedback = /Password must not contain/i;

        fireEvent.click(getByText("Wi-Fi 1"));

        const passwordInput = getByLabelText("Password");

        const changePassword = (value) =>
            fireEvent.change(passwordInput, { target: { value } });

        changePassword("12");
        expect(getByText(shortErrorFeedback)).toBeDefined();

        changePassword(
            "longpasswordlongpasswordlongpasswordlongpasswordlongpasswordlong"
        );
        expect(getByText(longErrorFeedback)).toBeDefined();
    });
});
