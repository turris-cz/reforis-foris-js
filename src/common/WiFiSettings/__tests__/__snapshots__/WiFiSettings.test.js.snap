// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WiFiSettings/> Snapshot 2.4 GHz 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -240,207 +240,95 @@
              value="0"
            >
              auto
            </option>
            <option
-             value="36"
+             value="1"
            >
              
-                         36
-                         (5180 MHz )
+                         1
+                         (2412 MHz )
                      
            </option>
            <option
-             value="40"
+             value="2"
            >
              
-                         40
-                         (5200 MHz )
+                         2
+                         (2417 MHz )
                      
            </option>
            <option
-             value="44"
+             value="3"
            >
              
-                         44
-                         (5220 MHz )
+                         3
+                         (2422 MHz )
                      
            </option>
            <option
-             value="48"
-           >
-             
-                         48
-                         (5240 MHz )
-                     
-           </option>
-           <option
-             value="52"
-           >
-             
-                         52
-                         (5260 MHz  ,DFS)
-                     
-           </option>
-           <option
-             value="56"
-           >
-             
-                         56
-                         (5280 MHz  ,DFS)
-                     
-           </option>
-           <option
-             value="60"
+             value="4"
            >
              
-                         60
-                         (5300 MHz  ,DFS)
+                         4
+                         (2427 MHz )
                      
            </option>
            <option
-             value="64"
+             value="5"
            >
              
-                         64
-                         (5320 MHz  ,DFS)
+                         5
+                         (2432 MHz )
                      
            </option>
            <option
-             value="100"
+             value="6"
            >
              
-                         100
-                         (5500 MHz  ,DFS)
+                         6
+                         (2437 MHz )
                      
            </option>
            <option
-             value="104"
+             value="7"
            >
              
-                         104
-                         (5520 MHz  ,DFS)
+                         7
+                         (2442 MHz )
                      
            </option>
            <option
-             value="108"
+             value="8"
            >
              
-                         108
-                         (5540 MHz  ,DFS)
+                         8
+                         (2447 MHz )
                      
            </option>
            <option
-             value="112"
+             value="9"
            >
              
-                         112
-                         (5560 MHz  ,DFS)
+                         9
+                         (2452 MHz )
                      
            </option>
            <option
-             value="116"
+             value="10"
            >
              
-                         116
-                         (5580 MHz  ,DFS)
+                         10
+                         (2457 MHz )
                      
            </option>
            <option
-             value="120"
+             value="11"
            >
              
-                         120
-                         (5600 MHz  ,DFS)
-                     
-           </option>
-           <option
-             value="124"
-           >
-             
-                         124
-                         (5620 MHz  ,DFS)
-                     
-           </option>
-           <option
-             value="128"
-           >
-             
-                         128
-                         (5640 MHz  ,DFS)
-                     
-           </option>
-           <option
-             value="132"
-           >
-             
-                         132
-                         (5660 MHz  ,DFS)
-                     
-           </option>
-           <option
-             value="136"
-           >
-             
-                         136
-                         (5680 MHz  ,DFS)
-                     
-           </option>
-           <option
-             value="140"
-           >
-             
-                         140
-                         (5700 MHz  ,DFS)
-                     
-           </option>
-           <option
-             value="144"
-           >
-             
-                         144
-                         (5720 MHz  ,DFS)
-                     
-           </option>
-           <option
-             value="149"
-           >
-             
-                         149
-                         (5745 MHz )
-                     
-           </option>
-           <option
-             value="153"
-           >
-             
-                         153
-                         (5765 MHz )
-                     
-           </option>
-           <option
-             value="157"
-           >
-             
-                         157
-                         (5785 MHz )
-                     
-           </option>
-           <option
-             value="161"
-           >
-             
-                         161
-                         (5805 MHz )
-                     
-           </option>
-           <option
-             value="165"
-           >
-             
-                         165
-                         (5825 MHz )
+                         11
+                         (2462 MHz )
                      
            </option>
          </select>
        </div>
        <div"
`;

exports[`<WiFiSettings/> Snapshot both modules disabled. 1`] = `
<DocumentFragment>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <form>
      <div
        class="form-check form-switch mb-3 d-flex align-items-center"
      >
        <input
          class="form-check-input me-2"
          id="1"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="1"
        >
          <h2
            class="mb-0"
          >
            Wi-Fi 1
          </h2>
        </label>
      </div>
      <hr />
      <div
        class="form-check form-switch mb-3 d-flex align-items-center"
      >
        <input
          class="form-check-input me-2"
          id="2"
          role="switch"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="2"
        >
          <h2
            class="mb-0"
          >
            Wi-Fi 2
          </h2>
        </label>
      </div>
      <div
        class="text-end"
      >
        <button
          class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
          type="submit"
        >
          <span>
            Save
          </span>
        </button>
      </div>
    </form>
  </div>
  <div
    class="card p-4 col-sm-12 col-lg-12 p-0 mb-4"
  >
    <h2>
      Reset Wi-Fi Settings
    </h2>
    <p>
      If a number of wireless cards doesn't match, you may try to reset the Wi-Fi settings. Note that this will remove the current Wi-Fi configuration and restore the default values.
    </p>
    <div
      class="text-end"
    >
      <button
        class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
        type="button"
      >
        <span>
          Reset Wi-Fi Settings
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<WiFiSettings/> Snapshot guest network. 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -524,10 +524,91 @@
            <small>
              Enables Wi-Fi for guests, which is separated from LAN network. Devices connected to this network are allowed to access the internet, but aren't allowed to access other devices and the configuration interface of the router. Parameters of the guest network can be set in the Guest network tab.
            </small>
          </div>
        </div>
+       <div
+         class="mb-3"
+       >
+         <label
+           class="form-label"
+           for="12"
+         >
+           SSID
+         </label>
+         <div
+           class="input-group"
+         >
+           <input
+             class="form-control"
+             id="12"
+             type="text"
+             value="TestGuestSSID"
+           />
+           <div
+             class="input-group-append"
+           >
+             <button
+               class="input-group-text"
+               type="button"
+             >
+               <i
+                 class="fa"
+               />
+             </button>
+           </div>
+         </div>
+         <div
+           class="form-text"
+         >
+           <small>
+             SSID which contains non-standard characters could cause problems on some devices.
+           </small>
+         </div>
+       </div>
+       <div
+         class="mb-3"
+       >
+         <label
+           class="form-label"
+           for="13"
+         >
+           Password
+         </label>
+         <div
+           class="input-group"
+         >
+           <input
+             autocomplete="current-password"
+             class="form-control is-invalid"
+             id="13"
+             required=""
+             type="password"
+             value=""
+           />
+           <button
+             class="input-group-text"
+             type="button"
+           >
+             <i
+               class="fa"
+             />
+           </button>
+         </div>
+         <div
+           class="invalid-feedback"
+         >
+           Password must contain at least 8 symbols
+         </div>
+         <div
+           class="form-text"
+         >
+           <small>
+             WPA2/3 pre-shared key, that is required to connect to the network.
+           </small>
+         </div>
+       </div>
        <hr />
        <div
          class="form-check form-switch mb-3 d-flex align-items-center"
        >
          <input
@@ -550,10 +631,11 @@
        <div
          class="text-end"
        >
          <button
            class="btn btn-primary col-12 col-md-3 col-lg-2 d-inline-flex justify-content-center align-items-center"
+           disabled=""
            type="submit"
          >
            <span>
              Save
            </span>"
`;

exports[`<WiFiSettings/> Snapshot one module enabled. 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -21,10 +21,513 @@
            >
              Wi-Fi 1
            </h2>
          </label>
        </div>
+       <div
+         class="mb-3"
+       >
+         <label
+           class="form-label"
+           for="3"
+         >
+           SSID
+         </label>
+         <div
+           class="input-group"
+         >
+           <input
+             class="form-control"
+             id="3"
+             required=""
+             type="text"
+             value="TestSSID1"
+           />
+           <button
+             class="input-group-text"
+             type="button"
+           >
+             <i
+               class="fa"
+             />
+           </button>
+         </div>
+         <div
+           class="form-text"
+         >
+           <small>
+             SSID which contains non-standard characters could cause problems on some devices.
+           </small>
+         </div>
+       </div>
+       <div
+         class="mb-3"
+       >
+         <label
+           class="form-label"
+           for="4"
+         >
+           Password
+         </label>
+         <div
+           class="input-group"
+         >
+           <input
+             autocomplete="current-password"
+             class="form-control"
+             id="4"
+             required=""
+             type="password"
+             value="TestPass"
+           />
+           <button
+             class="input-group-text"
+             type="button"
+           >
+             <i
+               class="fa"
+             />
+           </button>
+         </div>
+         <div
+           class="form-text"
+         >
+           <small>
+             WPA2/3 pre-shared key, that is required to connect to the network.
+           </small>
+         </div>
+       </div>
+       <div
+         class="form-check form-switch mb-3"
+       >
+         <input
+           class="form-check-input"
+           id="5"
+           role="switch"
+           type="checkbox"
+         />
+         <label
+           class="form-check-label"
+           for="5"
+         >
+           Hide SSID
+         </label>
+         <div
+           class="form-text"
+         >
+           <small>
+             If set, network is not visible when scanning for available networks.
+           </small>
+         </div>
+       </div>
+       <div
+         class="mb-3"
+       >
+         <label
+           class="d-block"
+           for="6"
+         >
+           GHz
+         </label>
+         <div
+           class="mb-3 form-check form-check-inline"
+         >
+           <input
+             class="form-check-input me-2"
+             id="hwmode-0-0"
+             name="hwmode-0"
+             type="radio"
+             value="11g"
+           />
+           <label
+             class="form-check-label"
+             for="hwmode-0-0"
+           >
+             2.4
+           </label>
+         </div>
+         <div
+           class="mb-3 form-check form-check-inline"
+         >
+           <input
+             checked=""
+             class="form-check-input me-2"
+             id="hwmode-0-1"
+             name="hwmode-0"
+             type="radio"
+             value="11a"
+           />
+           <label
+             class="form-check-label"
+             for="hwmode-0-1"
+           >
+             5
+           </label>
+         </div>
+         <div
+           class="form-text"
+         >
+           <small>
+             The 2.4 GHz band is more widely supported by clients, but tends to have more interference. The 5 GHz band is a newer standard and may not be supported by all your devices. It usually has less interference, but the signal does not carry so well indoors.
+           </small>
+         </div>
+       </div>
+       <div
+         class="mb-3"
+       >
+         <label
+           class="form-label"
+           for="7"
+         >
+           802.11n/ac/ax mode
+         </label>
+         <select
+           class="form-select"
+           id="7"
+         >
+           <option
+             value="NOHT"
+           >
+             Disabled
+           </option>
+           <option
+             value="HT20"
+           >
+             802.11n - 20 MHz wide channel
+           </option>
+           <option
+             value="HT40"
+           >
+             802.11n - 40 MHz wide channel
+           </option>
+           <option
+             value="VHT20"
+           >
+             802.11ac - 20 MHz wide channel
+           </option>
+           <option
+             value="VHT40"
+           >
+             802.11ac - 40 MHz wide channel
+           </option>
+           <option
+             value="VHT80"
+           >
+             802.11ac - 80 MHz wide channel
+           </option>
+         </select>
+         <div
+           class="form-text"
+         >
+           <small>
+             Change this to adjust 802.11n/ac/ax mode of operation. 802.11n with 40 MHz wide channels can yield higher throughput but can cause more interference in the network. If you don't know what to choose, use the default option with 20 MHz wide channel.
+           </small>
+         </div>
+       </div>
+       <div
+         class="mb-3"
+       >
+         <label
+           class="form-label"
+           for="8"
+         >
+           Channel
+         </label>
+         <select
+           class="form-select"
+           id="8"
+         >
+           <option
+             value="0"
+           >
+             auto
+           </option>
+           <option
+             value="36"
+           >
+             
+                         36
+                         (5180 MHz )
+                     
+           </option>
+           <option
+             value="40"
+           >
+             
+                         40
+                         (5200 MHz )
+                     
+           </option>
+           <option
+             value="44"
+           >
+             
+                         44
+                         (5220 MHz )
+                     
+           </option>
+           <option
+             value="48"
+           >
+             
+                         48
+                         (5240 MHz )
+                     
+           </option>
+           <option
+             value="52"
+           >
+             
+                         52
+                         (5260 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="56"
+           >
+             
+                         56
+                         (5280 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="60"
+           >
+             
+                         60
+                         (5300 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="64"
+           >
+             
+                         64
+                         (5320 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="100"
+           >
+             
+                         100
+                         (5500 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="104"
+           >
+             
+                         104
+                         (5520 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="108"
+           >
+             
+                         108
+                         (5540 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="112"
+           >
+             
+                         112
+                         (5560 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="116"
+           >
+             
+                         116
+                         (5580 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="120"
+           >
+             
+                         120
+                         (5600 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="124"
+           >
+             
+                         124
+                         (5620 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="128"
+           >
+             
+                         128
+                         (5640 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="132"
+           >
+             
+                         132
+                         (5660 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="136"
+           >
+             
+                         136
+                         (5680 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="140"
+           >
+             
+                         140
+                         (5700 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="144"
+           >
+             
+                         144
+                         (5720 MHz  ,DFS)
+                     
+           </option>
+           <option
+             value="149"
+           >
+             
+                         149
+                         (5745 MHz )
+                     
+           </option>
+           <option
+             value="153"
+           >
+             
+                         153
+                         (5765 MHz )
+                     
+           </option>
+           <option
+             value="157"
+           >
+             
+                         157
+                         (5785 MHz )
+                     
+           </option>
+           <option
+             value="161"
+           >
+             
+                         161
+                         (5805 MHz )
+                     
+           </option>
+           <option
+             value="165"
+           >
+             
+                         165
+                         (5825 MHz )
+                     
+           </option>
+         </select>
+       </div>
+       <div
+         class="mb-3"
+       >
+         <label
+           class="form-label"
+           for="9"
+         >
+           Encryption
+         </label>
+         <select
+           class="form-select"
+           id="9"
+         >
+           <option
+             value="WPA3"
+           >
+             WPA3 only
+           </option>
+           <option
+             value="WPA2/3"
+           >
+             WPA3 with WPA2 as fallback (default)
+           </option>
+           <option
+             value="WPA2"
+           >
+             WPA2 only
+           </option>
+         </select>
+         <div
+           class="form-text"
+         >
+           <small>
+             The WPA3 standard is the new most secure encryption method that is suggested to be used with any device that supports it. The older devices without WPA3 support require older WPA2. If you experience issues with connecting older devices, try to enable WPA2.
+           </small>
+         </div>
+       </div>
+       <div
+         class="form-check form-switch mb-3"
+       >
+         <input
+           class="form-check-input"
+           id="10"
+           role="switch"
+           type="checkbox"
+         />
+         <label
+           class="form-check-label"
+           for="10"
+         >
+           Disable Management Frame Protection
+         </label>
+         <div
+           class="form-text"
+         >
+           <small>
+             In case you have trouble connecting to WiFi Access Point, try disabling Management Frame Protection.
+           </small>
+         </div>
+       </div>
+       <div
+         class="form-check form-switch mb-3"
+       >
+         <input
+           class="form-check-input"
+           id="11"
+           role="switch"
+           type="checkbox"
+         />
+         <label
+           class="form-check-label"
+           for="11"
+         >
+           Enable Guest Wi-Fi
+         </label>
+         <div
+           class="form-text"
+         >
+           <small>
+             Enables Wi-Fi for guests, which is separated from LAN network. Devices connected to this network are allowed to access the internet, but aren't allowed to access other devices and the configuration interface of the router. Parameters of the guest network can be set in the Guest network tab.
+           </small>
+         </div>
+       </div>
        <hr />
        <div
          class="form-check form-switch mb-3 d-flex align-items-center"
        >
          <input"
`;
